server/datastore/mysql/mysql.go

```diff
@@ -477,10 +481,14 @@ func (d *Datastore) whereFilterHostsByTeams(filter fleet.TeamFilter, hostKey str
 
 	// Collect matching teams
 	var idStrs []string
+	var teamIDSeen bool
```

Comment:
So how I thought it'd fit is for when a team ID is explicitly provided, the user can see it if it could see any team (as seen with the `defaultAllowClause` above), or if the requested team ID is part of the team IDs that user can see (as implemented below). Otherwise nothing is returned. Let me know if it turns out that's not how the logic should be applied!